### PR TITLE
HOSTEDCP-839: Audit log sidecars for openshift-apiserver and openshift-oauth-apiserver

### PR DESCRIPTION
**Pull request by imain**
This adds a tail sidecar container for the audit logs for the above deployments. It
also plumbs the audit config into the two deployments.

Fixes #[HOSTEDCP-839](https://issues.redhat.com//browse/HOSTEDCP-839)

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go [hunk 1]
```diff
@@ -74,6 +74,27 @@ func ReconcileOAuthAPIServerDeployment(deployment *appsv1.Deployment, ownerRef c
...
+				},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("5m"),
+						corev1.ResourceMemory: resource.MustParse("10Mi"),
+					},
+				},
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      oauthVolumeWorkLogs().Name,
+					MountPath: "/var/log/openshift-apiserver",
```
Comment (imain) [inline control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go:95]: We're going to want our own container for this I presume?

**Comment by csrwng** [comment]
@imain Thank you this is awesome. We still need to configure the webhook for both apiservers in the same way we configure it for the Kube APIserver: 
https://github.com/openshift/hypershift/blob/5baf27ad8c671add9ae9d761515be43e1067fcd5/control-plane-operator/controllers/hostedcontrolplane/kas/config.go#L154-L157
https://github.com/openshift/hypershift/blob/5baf27ad8c671add9ae9d761515be43e1067fcd5/control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go#L742-L756

File: control-plane-operator/controllers/hostedcontrolplane/oapi/params.go [hunk 2]
```diff
@@ -189,8 +195,18 @@ func (p *OpenShiftAPIServerParams) IngressDomain() string {
 	return p.IngressSubDomain
 }
 
-func (p *OpenShiftAPIServerParams) OAuthAPIServerDeploymentParams() *OAuthDeploymentParams {
-	return &OAuthDeploymentParams{
+func (p *OpenShiftAPIServerParams) AuditPolicyConfig() configv1.Audit {
+	if p.APIServer != nil {
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/oapi/params.go:199]: please also check whether a profile is set
```go
if p.APIServer != nil && p.APIServer.Audit.Profile != "" {
```

File: control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go [hunk 3]
```diff
@@ -86,6 +107,13 @@ func ReconcileOAuthAPIServerDeployment(deployment *appsv1.Deployment, ownerRef c
 			util.BuildVolume(common.VolumeTotalClientCA(), common.BuildVolumeTotalClientCA),
 		},
 	}
+
+	/*
+		if auditWebhookRef != nil {
+			applyOASAuditWebhookConfigFileVolume(&deployment.Spec.Template.Spec, auditWebhookRef)
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go:113]: Looks like this should be uncommented

File: control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go [hunk 4]
```diff
@@ -74,6 +74,27 @@ func ReconcileOAuthAPIServerDeployment(deployment *appsv1.Deployment, ownerRef c
...
+				},
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("5m"),
+						corev1.ResourceMemory: resource.MustParse("10Mi"),
+					},
+				},
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      oauthVolumeWorkLogs().Name,
+					MountPath: "/var/log/openshift-oauth-apiserver",
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go:95]: pls use util.PodVolumeMounts as with the main container. The reason to do that is that if in the future we change the path, we don't have to look multiple places to fix it.

File: control-plane-operator/controllers/hostedcontrolplane/oapi/deployment.go [hunk 5]
```diff
@@ -109,6 +115,27 @@ func ReconcileDeployment(deployment *appsv1.Deployment, ownerRef config.OwnerRef
...
 			util.BuildContainer(oasContainerMain(), buildOASContainerMain(image, strings.Split(etcdUrlData.Host, ":")[0], defaultOAPIPort)),
+			{
+				Name:            "audit-logs",
+				Image:           image,
+				ImagePullPolicy: corev1.PullIfNotPresent,
+				Command: []string{
+					"/usr/bin/tail",
+					"-c+1",
+					"-F",
+					"/var/log/openshift-apiserver/audit.log",
```
Comment (csrwng) [inline control-plane-operator/controllers/hostedcontrolplane/oapi/deployment.go:126]: pls use util.PodVolumeMounts

**Review by csrwng — COMMENTED** [review]
@imain left some comments. Thx!

File: control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go [hunk 6]
```diff
@@ -155,7 +155,7 @@ func ReconcileKubeAPIServerDeployment(deployment *appsv1.Deployment,
 						"/usr/bin/tail",
 						"-c+1",
 						"-F",
-						"/var/log/kube-apiserver/audit.log",
+						fmt.Sprintf("%s/%s", volumeMounts.Path(kasContainerMain().Name, kasVolumeWorkLogs().Name), "audit.log"),
```
Comment (imain) [inline control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go:158]: Is this what you were thinking @csrwng ?

Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go:158]: Yes, thank you. The reason is that if you change the volume path in one place, you don't have to worry about changing it everywhere else it's referenced.

**Comment by csrwng** [comment]
/lgtm
@imain pls remove hold when you are done with your testing, thx!

**Comment by imain** [comment]
OK, tested it!  Seems to be working!  I did however have a hard time distinguishing which service was sending the message..

**Comment by csrwng** [comment]
@imain it looks like the control-plane-operator is crashing with:
```
panic: invalid pod container oauth-apiserver used when looking for mount [recovered]
	panic: invalid pod container oauth-apiserver used when looking for mount

goroutine 63 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/hypershift/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:118 +0x2d9
panic({0x33329a0, 0xc00174f100})
	/usr/lib/golang/src/runtime/panic.go:890 +0x267
github.com/openshift/hypershift/support/util.PodVolumeMounts.Path(0xc000c6f260, {0x39cf81e, 0xf}, {0x39c175e, 0x9})
	/hypershift/support/util/volumemounts.go:16 +0x34e
github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/oapi.ReconcileOAuthAPIServerDeployment(0xc001732d80, {0xc000833a90}, 0xc00173a900, 0xc000b5e168)
	/hypershift/control-plane-operator/controllers/hostedcontrolplane/oapi/oauth_deployment.go:90 +0x4eb
```

**Comment by enxebre** [comment]
Thanks!
/lgtm
/cherry-pick release-4.13
/cherry-pick release-4.12

**Pull request merged**